### DISCUSSION TITLE
Maven is extremely slow

### DISCUSSION THREAD

**Discussion by TheSuperEngineer** [checkstyle/checkstyle#12895] [General]
As a regular and dedicated contributor I would like to point out that running ``mvn clean verify`` takes eternity to complete. I have to run ``mvn clean verify`` several times so that the CI resources are not unnecessarily consumed when I make a PR. Is there a faster way to complete ``mvn clean verify``? Note that I am using windows 11.
(GPTs solution)
![image](https://user-images.githubusercontent.com/90137881/227340764-36078a65-cb39-4645-8d3e-b970745c4c6f.png)

**Comment by nrmancuso** [comment]
@shubh220922 it will help to become familiar with mvn CLI, so that you can run only related steps until you think you are ready to push, then run the full `mvn clean verify`. I usually just run related tests and ant-phase-verify until I think I am ready.

How long is mvn  clean verify taking for you?

**Reply by TheSuperEngineer** [reply]
It ranges from as less as 7 minutes to as long as 14 minutes. On linux it is faster while on windows it happens to be slower on average though this can be very specific for me.

**Reply by TheSuperEngineer** [reply]
@nrmancuso you are right. It is a great way to learn to use maven in a much more efficient manner. I'll run selective tests from now on.

**Reply by nrmancuso** [reply]
@shubh220922 you can try to give it more memory, you can google this and find some stackoverflow posts that might help.

**Reply by TheSuperEngineer** [reply]
ok let me see this. Work is required at my end.

**Reply by rnveach** [reply]
Giving more memory won't help unless your computer is already slow or slow with 2 gigs of RAM gone. CS cannot run parallel tests, especially with issues like the locale.

Your best bet, as mentioned, is learn the maven lifecycles and the CS lifecycles and know what to trim back on runs for faster assurance. The full run will still take time, but what you change may not impact everything. You are not speeding up the run. Instead you are just doing less checks.

One example is `clean`. You don't always need to do it, and some places we have caches that speed up processing on multiple runs. Except those don't help when you clean them away. However, if you are not careful and never clean, you can have weird issues crop up.

Even on my beefy laptop, the full run takes 9.2 minutes.

**Comment by romani** [comment]
If you need to resolve certain validation , run certain validation separately https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations

Do just verify only at the most end.
Before push recheck spelling by script execution.

**Reply by TheSuperEngineer** [reply]
Thanks a lot for sharing. I wasted too much time checking for PMD violation and sevntu checks by running ``mvn clean verify`` again and again.
